transport/httptransport/http_transport.go

```diff
@@ -0,0 +1,241 @@
+	// init the request with the transfer context
+	req = req.WithContext(ctx)
+	// open output file in append-only mode for writing
+	of, err := os.OpenFile(t.dealInfo.OutputFile, os.O_APPEND|os.O_WRONLY, 0644)
+	if err != nil {
+		failTransfer(fmt.Errorf("failed to open output file: %w", err))
+		return
+	}
+
+	// start the http transfer
+	remaining := t.dealInfo.DealSize - fi.Size()
+	if err := t.doHttp(req, of, remaining); err != nil {
+		// do not resume transfer if context has been cancelled
+		if ctx.Err() == context.Canceled || ctx.Err() == context.DeadlineExceeded {
+			if err := t.pub.Emit(types.TransportEvent{
+				Status:          transferstatus.Cancelled,
+				IsTerminalState: true,
+				Error:           fmt.Errorf("transfer context err: %w", ctx.Err()),
+			}); err != nil {
+				log.Errorw("failed to publish transport error", "dealUuid", t.dealInfo.DealUuid, "err", err)
+			}
+			return
+		}
+
+		// TODO: resumption
+		return
+	}
+
+	// --- http request finished successfully. see if we got the number of bytes we expected.
+
+	// if the number of bytes we've received is not the same as the deal size, we have a failure.
+	if t.nBytesReceived != t.dealInfo.DealSize {
+		failTransfer(fmt.Errorf("mismatch in dealSize vs received bytes, dealSize=%d, received=%d", t.dealInfo.DealSize, t.nBytesReceived))
+		return
+	}
+
+	// otherwise we're good ! notify the subscriber.
+	if err := t.pub.Emit(types.TransportEvent{
+		Status:          transferstatus.Finished,
+		NBytesReceived:  t.nBytesReceived,
+		IsTerminalState: true,
+	}); err != nil {
+		log.Errorw("failed to publish transport event", "dealUuid", t.dealInfo.DealUuid, "err", err)
+	}
+}
+
+func (t *transfer) doHttp(req *http.Request, dst io.Writer, toRead int64) error {
+	// send http request and validate response
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send  http req: %w", err)
+	}
+	// we should either get back a 200 or a 206 -> anything else means something has gone wrong and we return an error.
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusPartialContent {
+		resp.Body.Close() // nolint
+		return fmt.Errorf("http req failed: %d", resp.StatusCode)
+	}
+	defer resp.Body.Close()
+
+	//  start reading the response stream `readBufferSize` at a time using a limit reader so we only read as many bytes as we need to.
+	buf := make([]byte, readBufferSize)
+	limitR := io.LimitReader(resp.Body, toRead)
+	for {
+		nr, readErr := limitR.Read(buf)
+
+		// if we read more than zero bytes, write whatever read.
+		if nr > 0 {
+			nw, writeErr := dst.Write(buf[0:nr])
+
+			// if the number of read and written bytes don't match -> something has gone wrong, abort the http req.
+			if nw < 0 || nr != nw {
+				if writeErr != nil {
+					return fmt.Errorf("failed to write to output file: %w", err)
+				}
+				return fmt.Errorf("read-write mismatch, read=%d, written=%d", nr, nw)
```

Comment:
nit: make clear this error arises when writing the file (not when reading from the HTTP stream)